src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+
+void string_refine_preprocesst::make_to_char_array_function(
+  goto_programt &goto_program, goto_programt::targett &i_it)
+{
+  const code_function_callt &function_call=to_code_function_call(i_it->code);
+
+  assert(function_call.arguments().size()>=1);
+  const exprt &string_argument=function_call.arguments()[0];
+  assert(refined_string_typet::is_java_string_pointer_type(
+           string_argument.type()));
+
+  dereference_exprt deref(string_argument, string_argument.type().subtype());
+  typet length_type, data_type;
+  get_data_and_length_type_of_string(deref, data_type, length_type);
+  member_exprt rhs_data(deref, "data", data_type);
+  member_exprt rhs_length(deref, "length", length_type);
+  dereference_exprt rhs_array(rhs_data, data_type.subtype());
+  exprt first_index=from_integer(0, java_int_type());
+  index_exprt first_element(rhs_array, first_index, java_char_type());
+  address_of_exprt rhs_pointer(first_element);
+  typet deref_type=function_call.lhs().type().subtype();
+  dereference_exprt deref_lhs(function_call.lhs(), deref_type);
+  member_exprt lhs_data(deref_lhs, "data", data_type);
+  member_exprt lhs_length(deref_lhs, "length", length_type);
+  std::list<code_assignt> assignments;
+  assignments.emplace_back(lhs_data, rhs_pointer);
+  assignments.emplace_back(lhs_length, rhs_length);
+  insert_assignments(goto_program, i_it, assignments);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_cprover_char_array_assign
+
+  Inputs: a goto_program, a position in this program, an expression and a
+          location
+
+ Outputs: a char array expression (not a pointer)
+
+ Purpose: Introduce a temporary variable for cprover strings;
+          returns the cprover_string corresponding to rhs
+
+\*******************************************************************/
+
+exprt string_refine_preprocesst::make_cprover_char_array_assign(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const exprt &rhs,
+  const source_locationt &location)
+{
+  // TODO : add an assertion on the type of rhs
+
+  // We do the following assignments:
+  // cprover_string_array = rhs.data
+  // cprover_string = { rhs.length; cprover_string_array }
+
+  // string expression for the rhs of the second assignment
+  string_exprt new_rhs(java_char_type());
+
+  typet data_type=new_rhs.content().type();
+  typet length_type=java_int_type();
+
+  symbol_exprt array_lhs=new_symbol("cprover_string_array", data_type);
+  member_exprt array_rhs(rhs, "data", new_rhs.content().type());
+  symbol_exprt lhs=new_symbol("cprover_string", new_rhs.type());
+  new_rhs.length()=member_exprt(rhs, "length", length_type);
+  new_rhs.content()=array_lhs;
+
+  std::list<code_assignt> assignments;
+  assignments.emplace_back(array_lhs, array_rhs);
+  assignments.emplace_back(lhs, new_rhs);
+  insert_assignments(goto_program, i_it, assignments);
```

Comment:
In that case, we could write something like `insert_assignments(goto_program, i_it, {code_assignt(array_lhs, array_rhs), code_assignt(lhs, new_rhs)});` but it will lose a bit of the structure introduced in https://github.com/diffblue/cbmc/pull/378/commits/979ec9469fa90726f3122e8a64a3170319755a9c